discovery-provider/src/tasks/index_spl_token.py

```diff
@@ -0,0 +1,451 @@
+import json
+import logging
+import time
+from typing import List, Optional, Set, Tuple
+
+import base58
+from redis import Redis
+from solana.publickey import PublicKey
+from src.models import (
+    AssociatedWallet,
+    SPLTokenTransaction,
+    User,
+    UserBankAccount,
+    WalletChain,
+)
+from src.queries.get_balances import enqueue_immediate_balance_refresh
+from src.solana.constants import (
+    TX_SIGNATURES_BATCH_SIZE,
+    TX_SIGNATURES_MAX_BATCHES,
+    TX_SIGNATURES_RESIZE_LENGTH,
+)
+from src.solana.solana_client_manager import SolanaClientManager
+from src.solana.solana_helpers import get_base_address
+from src.solana.solana_transaction_types import (
+    ConfirmedSignatureForAddressResult,
+    ConfirmedTransaction,
+    TransactionInfoResult,
+)
+from src.tasks.celery_app import celery
+from src.utils.cache_solana_program import (
+    CachedProgramTxInfo,
+    cache_latest_sol_db_tx,
+    fetch_and_cache_latest_program_tx_redis,
+)
+from src.utils.config import shared_config
+from src.utils.redis_constants import (
+    latest_sol_spl_token_db_key,
+    latest_sol_spl_token_program_tx_key,
+)
+from src.utils.session_manager import SessionManager
+from src.utils.solana_indexing_logger import SolanaIndexingLogger
+
+SPL_TOKEN_PROGRAM = shared_config["solana"]["waudio_mint"]
+SPL_TOKEN_PUBKEY = PublicKey(SPL_TOKEN_PROGRAM) if SPL_TOKEN_PROGRAM else None
+USER_BANK_ADDRESS = shared_config["solana"]["user_bank_program_address"]
+USER_BANK_PUBKEY = PublicKey(USER_BANK_ADDRESS) if USER_BANK_ADDRESS else None
+
+REDIS_TX_CACHE_QUEUE_PREFIX = "spl-token-tx-cache-queue"
+
+# Number of signatures that are fetched from RPC and written at once
+# For example, in a batch of 1000 only 100 will be fetched and written in parallel
+# Intended to relieve RPC and DB pressure
+TX_SIGNATURES_PROCESSING_SIZE = 100
+
+logger = logging.getLogger(__name__)
+
+
+def get_token_balance_change_owners(
+    tx_result: TransactionInfoResult,
+) -> Tuple[Set[str], Set[str]]:
+    root_accounts_to_refresh: Set[str] = set()
+    token_accounts_to_refresh: Set[str] = set()
+    if (
+        "meta" not in tx_result
+        or "preTokenBalances" not in tx_result["meta"]
+        or "postTokenBalances" not in tx_result["meta"]
+        or "transaction" not in tx_result
+        or "message" not in tx_result["transaction"]
+        or "accountKeys" not in tx_result["transaction"]["message"]
+    ):
+        logger.info("iomnvalid format, return early")
+        return (root_accounts_to_refresh, token_accounts_to_refresh)
+    owner_balance_dict = {}
+    account_keys = tx_result["transaction"]["message"]["accountKeys"]
+    for pre_balance in tx_result["meta"]["preTokenBalances"]:
+        owner_balance_dict[pre_balance["accountIndex"]] = pre_balance["uiTokenAmount"][
+            "amount"
+        ]
+    for post_balance in tx_result["meta"]["postTokenBalances"]:
+        if (
+            post_balance["accountIndex"] not in owner_balance_dict
```

Comment:
Can we pull out some temp vars for things like post_balance["account_index"], this is pretty confusing to follow